sql/core/src/main/scala/org/apache/spark/sql/execution/aggregate/HashAggregateExec.scala

```diff
@@ -847,18 +904,60 @@ case class HashAggregateExec(
       s"UnsafeRow $unsafeRowBuffer = null;"
     }
 
+    val adaptivePartialAggregation = if (isAdaptivePartialAggregationEnabled) {
+      val numberOfConsumedKeysTerm =
+        ctx.addMutableState(CodeGenerator.JAVA_LONG, "numberOfConsumedKeys")
+      val numAggSkippedRows = metricTerm(ctx, "numSkippedParAggRows")
+      val initExpr = declFunctions.flatMap(f => f.initialValues)
+      val emptyBufferKeyCode = GenerateUnsafeProjection.createCode(ctx, initExpr)
+
+      val numberOfKeys = if (fastHashMapTerm != null) {
+        s"$numberOfConsumedKeysTerm = $fastHashMapTerm.getNumKeys();"
+      } else if (hashMapTerm != null) {
+        s"$numberOfConsumedKeysTerm = $hashMapTerm.getNumKeys();"
```

Comment:
When fastHashMap is enabled, this implementation doesn’t count keys in regular hashMap. However, even when number of processed keys is less than fastHashMap capacity, the regularHashMap seems can still be non-empty when the fastHashMap
        1. unable to find group key within maxSteps (maxSteps = 2)
        2. run out of max supported page size. 

So even though adaptivePartialAggregationThreshold must be smaller than fastHashMap capacity, it's still possible that both fastHashMap and regular hashMap are non-empty. A potential fix can be `numberOfKeys = fastHashMap.getNumKeys() + hashMap.getNumKeys()`, and don't enable skipPartialAggregate once the aggregate spills. With this fix we can also remove [the adaptivePartialAggregationThreshold < fastHashMap's capacity check](https://github.com/apache/spark/pull/35806/files#diff-13c5b65678b327277c68d17910ae93629801af00117a0e3da007afd95b6c6764R2917-R2919)